specs/spec.md

```diff
@@ -0,0 +1,329 @@
+# @TODO
+- [ ] Link to parsers.
+- [ ] attestations
+- [ ] what else needs to be defined in the transport header?
+
+# Hobbits
+
+*Written by Dean Eigenmann <dean@status.im> & Rene Nayman <rene@whiteblock.io>* 
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Flow](#flow)
+3. [Wire Protocol](#wire-protocol)
+    - [Messages](#messages)
+4. [Transport](#transport)
+5. [Protocols](#protocols)
+    - [RPC](#rpc)
+    - [Gossip](#gossip)
+    - [PING](#ping)
+6. [Implementations](#implementations)
+
+## Abstract
+
+Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.
+
+Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.
+
+## Flow
+
+We layout a network where the nodes are peered in the following manner:
+
+![nodes](./nodes.png)
+
+The following flow dictates how blocks are shared amongst peers:
+
+![sequence diagram](./diagram.png)
+
+1. Node 1 receives a new block `0xCAB` and gossips `hash(0xCAB)` to Node 2 and Node 3
+2. Node 2 requests the block bodies for block `0xCAB` from Node 1 in an RPC Command query
+3. Node 1 responds to Node 2 with the block bodies for block `0xCAB` in an RPC Command response
+4. Node 2 receives a new block `0xCAB` and gossips it to Node 1 and Node 3
+
+## Wire Protocol
+
+### Messages
+
+The `message` format looks as follows:
+
+```
+EWP <version> <protocol> <header-length> <body-length>
+<header><body>
+```
+
+**NOTE:** Message`body`s along with `header` fields are BSON encoded.
```

Comment:
Is this true?  I mean, required?  The wire protocol just says, arbitrary binary data, for the <header> and <body>; and there's header-length and body-length to tell how to extract it.  

Maybe the BSON-encoding might happen in a layer above the wire-protocol, in GOSSIP/etc?  I'm not sure that it's required here.  

Similar comment about the "messages are utf-8 encoded strings" line, too... that would rule out binary payloads unless they're encoded first

Reply:
That’s how it was before. But we’re trying to simplify and cram all the details into a minimum doc so we can get a bare minimum interop going.